Make agents wait for dev_set_config/WPS PBC before master discovery

vitalii-komisarenko opened the pull request:
https://jira.prplfoundation.org/browse/PPM-143

In certification mode the agent state machine should stop before the MASTER_DISCOVERY state until dev_set_config is received (wired backhaul) or start_wps_registration (wireless backhaul). 

This PR should be merged after PR #1590.

vitalii-komisarenko committed:
agent: backhaul_manager: wait for dev_set_config in certification

https://jira.prplfoundation.org/browse/PPM-143

In certification mode the state machine of backhaul manager should stop
before entering MASTER_DISCOVERY state until dev_set_config (wired backhaul)
or start_wps_registration (wireless backhaul) is received.

No code change needed for wireless flow.

In case of wired flow the exception is made for the agent running on the same
machine as controller. In that case UCC listener is present only on the controller,
so there is no easy way to send dev_set_config to the agent.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

ghost commented:
@vitalii-komisarenko please also check to see if there is any documentation about the state machine.
If it exists -> update it
Otherwise, it could be good to add it

itayx commented:
Log entries taken from the agent logs (Filtered to show only the backhaul_manager_thread FSM prints)

From the master branch
```
TRACE 10:23:42:970 <140023053342464> backhaul_manager_thread.cpp[795] --> FSM: INIT --> WAIT_ENABLE
TRACE 10:23:46:045 <140023053342464> backhaul_manager_thread.cpp[809] --> FSM: WAIT_ENABLE --> ENABLED
TRACE 10:23:46:546 <140023053342464> backhaul_manager_thread.cpp[900] --> FSM: ENABLED --> MASTER_DISCOVERY
TRACE 10:23:46:547 <140023053342464> backhaul_manager_thread.cpp[935] --> FSM: MASTER_DISCOVERY --> SEND_AUTOCONFIG_SEARCH_MESSAGE
TRACE 10:23:47:055 <140023053342464> backhaul_manager_thread.cpp[973] --> FSM: SEND_AUTOCONFIG_SEARCH_MESSAGE --> WAIT_FOR_AUTOCONFIG_RESPONSE_MESSAGE
TRACE 10:23:47:058 <140023053342464> backhaul_manager_thread.cpp[3585] --> FSM: WAIT_FOR_AUTOCONFIG_RESPONSE_MESSAGE --> CONNECT_TO_MASTER
TRACE 10:23:47:058 <140023053342464> backhaul_manager_thread.cpp[983] --> FSM: CONNECT_TO_MASTER --> CONNECTED
TRACE 10:23:47:066 <140023053342464> backhaul_manager_thread.cpp[998] --> FSM: CONNECTED --> OPERATIONAL
```
From this PR
```
TRACE 10:26:13:974 <140670739035904> backhaul_manager_thread.cpp[795] --> FSM: INIT --> WAIT_ENABLE
TRACE 10:26:17:060 <140670739035904> backhaul_manager_thread.cpp[809] --> FSM: WAIT_ENABLE --> ENABLED
TRACE 10:26:17:561 <140670739035904> backhaul_manager_thread.cpp[900] --> FSM: ENABLED --> READY_FOR_MASTER_DISCOVERY
ERROR 11:08:03:977 <140671996204800> backhaul_manager_thread.cpp[921] --> m_agent_ucc_listener == nullptr
```
It looks as if the FSM goes to state `READY_FOR_MASTER_DISCOVERY` then fails because the listener is `null`

vitalii-komisarenko commented:
@itayx it is OK if the backhaul is wireless. The change is working only for the wired ones.

orenvor committed:
agent: Move front radio configuration to AgentDB

Move front radio configuration from "sPlatformSettings" struct
to the AgentDB.
Replace the use of front radio settings on the Agent thread to the one
on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move back radio configuration to AgentDB

Move back radio configuration from "sPlatformSettings" struct
to the AgentDB.
Replace the use of back radio settings on the Agent thread to the one
on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move mem_only_psk configuration to AgentDB

Move "mem_only_psk" configuration of back radio from "sPlatformSettings"
struct to the AgentDB.
Replace the use of "mem_only_psk" settings on the Agent thread to
the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move backhaul_max_vaps configuration to AgentDB

Move "backhaul_max_vaps" configuration of back radio from
"sPlatformSettings" struct to the AgentDB.
Replace the use of "backhaul_max_vaps" settings on the Agent thread to
the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move backhaul_network_enabled configuration to AgentDB

Move "backhaul_network_enabled" configuration of back radio from
"sPlatformSettings" struct to the AgentDB.
Replace the use of "backhaul_network_enabled" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move backhaul_preferred_radio_band configuration to AgentDB

Move "backhaul_preferred_radio_band" configuration of back radio from
"sPlatformSettings" struct to the AgentDB.
Replace the use of "backhaul_preferred_radio_band" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move operating_mode configuration to AgentDB

Move "operating_mode" configuration of from "sPlatformSettings" struct
to the AgentDB.
Replace the use of "operating_mode" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move management_mode configuration to AgentDB

Move "management_mode"  configuration of from "sPlatformSettings" struct
to the AgentDB.
Replace the use of "management_mode" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move certification_mode configuration to AgentDB

Move "certification_mode" configuration of from "sPlatformSettings"
struct to the AgentDB.
Replace the use of "certification_mode" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move stop_on_failure_attempts configuration to AgentDB

Move "stop_on_failure_attempts" configuration from
"sPlatformSettings" struct to the AgentDB.
Replace the use of "stop_on_failure_attempts" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move client_band_steering_enabled configuration to AgentDB

Move "client_band_steering_enabled" configuration from
"sPlatformSettings" struct to the AgentDB.
Replace the use of "client_band_steering_enabled" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move client_optimal_path_roaming_enabled configuration to AgentDB

Move "client_optimal_path_roaming_enabled" configuration from
"sPlatformSettings" struct to the AgentDB.
Replace the use of "client_optimal_path_roaming_enabled" settings on the
Agent thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move dfs_reentry_enabled configuration to AgentDB

Move "dfs_reentry_enabled" configuration from "sPlatformSettings" struct
to the AgentDB.
Replace the use of "dfs_reentry_enabled" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move client_optimal_path_prefer_signal_strength to AgentDB

Move "client_optimal_path_roaming_prefer_signal_strength_enabled"
configuration from "sPlatformSettings" struct to the AgentDB.
Replace the use of
"client_optimal_path_roaming_prefer_signal_strength_enabled"
settings on the Agent thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move client_11k_roaming_enabled configuration to AgentDB

Move "client_11k_roaming_enabled" configuration from "sPlatformSettings"
struct to the AgentDB.
Replace the use of "client_11k_roaming_enabled" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move load_balancing_enabled configuration to AgentDB

Move "load_balancing_enabled" configuration from "sPlatformSettings"
struct to the AgentDB.
Replace the use of "load_balancing_enabled" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move service_fairness_enabled configuration to AgentDB

Move "service_fairness_enabled" configuration from "sPlatformSettings"
struct to the AgentDB.
Replace the use of "service_fairness_enabled" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Move rdkb_extensions_enabled configuration to AgentDB

Move "rdkb_extensions_enabled" configuration of back radio from
"sPlatformSettings" struct to the AgentDB.
Replace the use of "rdkb_extensions_enabled" settings on the Agent
thread to the one on the AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
tlvf: beerocks_message::sPlatformSettings remove unused settings

Remove of unused fields from sPlatformSettings struct.
Those fields have been moved to the Agent_DB and there for no longer
needed as part of a beerocks massage.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

vitalii-komisarenko committed:
test_flows.py: add timeouts to make tests pass

Synchronization in test_flows.py is implemented by using timeouts.

They are too short to pass tests on my personal computer,
but they occasionally fail on the CI too.

Added more timeouts to make tests more robust.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
test_gw_repeater: disable operational test for repeaters

https://jira.prplfoundation.org/browse/PPM-143

Disable operational test for repeaters during environmental docker launch.

In certification mode we want to change agents logic to wait for
dev_set_config or WPS PBC in order to proceed to master discovery.

In environmental docker launch subroutine the check for operational state
is done right after launching the containers.

One of the options is changing `test_gw_repeater.sh` in order to send
dev_set_config there, but for sake of simplicity it was chosen to send
dev_set_config in `test_flows.py` (will be introduced in a later commit in this PR).

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
test_flows: send dev_set_config to repeaters before tests

https://jira.prplfoundation.org/browse/PPM-143

Send dev_set_config to the repeaters as a preparation to
changing agents logic to wait to dev_set_config or WPS PBC
before going to master discovery in certification mode.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

orenvor committed:
agent: remove unused onboarding configuration

Remove unused "onboarding" configuration from "sPlatformSettings"
struct.

Replace the use of "onboarding" with false value.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
agent: Remove "platform_settings" local variable from son_slave_thread

Remove "platform_settings" local variable from son_slave_thread.
Replace the use of "platform_settings" to the settings in AgentDB.

Removing of the config from the tlvf will take place on future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

morantr commented on tests/test_flows.py:
```diff
@@ -317,7 +317,7 @@ def test_dev_reset_default(self):
         env.checkpoint()
         agent.cmd_reply("dev_set_config,backhaul,eth")
-        time.sleep(1)
+        time.sleep(3)
```
I did similar change on my PR.
https://github.com/prplfoundation/prplMesh/pull/1593/commits/578a2141ca98832e595d1e2570f904e519272f35

vitalii-komisarenko replied:
Apparently your hardware is better than mine. I need two extra seconds and you need only one :)

mariomaz replied:
This is a recurrent problem. Tests that rely on sleeps to pass are very brittle.

`check_cmdu_type` should have a timeout parameter and do polling until condition is satisfied or timeout expires. No need for sleep then.

morantr commented on tests/test_flows.py:
```diff
@@ -1327,5 +1327,11 @@ def test_beacon_report_query(self):
+        debug("sending dev_set_config to an agent")
+        agent.cmd_reply("dev_set_config,program,map,backhaul,eth")
+
+    time.sleep(7)
```
Why 7 seconds and not less?

vitalii-komisarenko replied:
Otherwise it fails on my laptop :)

mariomaz replied:
Too brittle, IMO

vitalii-komisarenko replied:
Yes, it is brittle, but it is out of scope of this PR. PPM-354 is created for this purpose.

morantr commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -897,13 +897,36 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
+            FSM_MOVE_STATE(MASTER_DISCOVERY);
+        }
+        break;
+    }
```
Well, there is no really good reason to create a whole state for this.
You could simply do the whole thing on the "MASTER_DISCOVERY" state, and if the requirements are not met, just break, and the next time that the state will run it will try again.

```cpp
    case EState::MASTER_DISCOVERY: {
        ### After PR #1590 will be merged please take this flag from the AgentDB
        auto certification_mode = bpl::cfg_get_certification_mode();
        if (certification_mode < 0) {
            LOG(ERROR) << "Failed reading certification_mode: " << certification_mode;
            return false;
        }

        // In certification mode we want to wait till dev_set_config is received (wired backhaul)
        // or start_wps_registration (wireless backhaul).
        if (certification_mode && m_agent_ucc_listener && !m_agent_ucc_listener->has_received_dev_set_config()) {
            break;
        }

        // Rest of MASTER_DISCOVERY state...
    }
```

vitalii-komisarenko replied:
Unfortunately it is not enough.
Even in certification mode dev_set_config is not always required.
We don't need it if WPS PBC is used or if the agent is run on the same machine where controller is run (we discussed `if (db->device_conf.local_controller && db->device_conf.local_gw) {` shortcut in the slack).

We still need an additional piece of data to check whether waiting for dev_set_config is required, and IMO an explicit state is better than a variable that acts like a state.

morantr replied:
You can add whatever validators you think we need. But you should remove the new state. 
A state of the backhaul manager should reflect a real logical state and not some specific scenario that is relevant only for certification.
If you want you can wrap it in a function and according to the function return value, decide if you continue with the state or break. That is also an option.
Either way, the state should be removed.

mariomaz replied:
I'm too late here but I share @vitalii-komisarenko's opinion on this matter: despite required functionality can be achieved with some state variables, it deserves a new state (and the decision should not be taken based on if this new state is reached only in certification).

morantr reviewed (changes requested):
Please remove the added state and I will approve :)

orenvor committed:
agent: Move wlan_settings configuration to AgentDB

Move "wlan_settings" configuration from son_slave_thread local variable
to the AgentDB.

Currently son_slave receive "wlan_settings" from platform_manager
using "PLATFORM_SON_SLAVE_REGISTER_REQUEST/RESPONSE" messages,
and save the settings in a local variable.

This transfer will be replaced by the use of agent_db to pass and
store "wlan_settings".

Removing of "sWlanSettings" from the tlvf will take place on
future commit.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
fixup! agent: Move wlan_settings configuration to AgentDB

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

orenvor committed:
tlvf: PLATFORM_SON_SLAVE_REGISTER_RESPONSE remove unused fields

Remove of unused fields: sPlatformSettings and wlan_settings
from PLATFORM_SON_SLAVE_REGISTER_RESPONSE message.
Those fields have been moved to the Agent_DB and there for no longer
needed as part of this beerocks massage.

PPM-338

Signed-off-by: Oren Vormaser <oren.vormaser@devalore.com>

morantr commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4125,4 +4125,9 @@ std::shared_ptr<bwl::sta_wlan_hal> backhaul_manager::get_selected_backhaul_sta_w
+bool backhaul_manager::is_main_agent() {
+    auto db = AgentDB::get();
+    return db->device_conf.local_controller && db->device_conf.local_gw;
+}
```
IMO this function is redundant. 
It is more clear to see the flags directly, and a logical AND of two flags does not justify creating a function.
Also "main_agent" is a little vague and confusing. For example, we use the term "main agent" to describe the process that contains the backhaul manager and the platform manager.
I really think that this commit should be dropped.

morantr commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -792,8 +792,33 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
+
+        bool wired_backhaul = db->backhaul.connection_type == AgentDB::sBackhaul::eConnectionType::Wired;
+
+        if (certification_mode && wired_backhaul && !is_main_agent()) {
```
You defined "main_agent" as local_gw && local_controller. Why is relevant that it will not be gw? 
AFAIU it should be:
`if (certification_mode && wired_backhaul && !db->devic_conf.local_controller)`
The controller does not neccesarely have to run on the GW.
This just enforces what I said on the new `main_agent` function.

morantr replied:
Please rebase to #1590 and take the `certification_mode` directly from the DB. It will save the error handling and you could use it directly on the condition.

vitalii-komisarenko commented:
This PR uses changes from PR #1590.

morantr commented:
Approved just make sure not to merge it until #1590 is merged.

mariomaz commented on tests/test_gw_repeater.sh:
```diff
@@ -128,18 +128,13 @@ main() {
-    }
+    # Testing for operational status of the repeaters is disabled in PR #1572.
+    # In certification mode agents shall wait for dev_set_config or WPS PBC,
+    # so they are inoperational at this stage anyway.
```
A trade-off solution would have been to redefine the concept of "operational". Some other check should be performed to know if agent is up.

vitalii-komisarenko replied:
Well, we still check the gateway... 
If I am not mistaken, it is used only by `test_flow.py`. If something goes wrong, the tests will fail.
But yes, it is not ideal.

mariomaz commented on agent/src/beerocks/slave/agent_ucc_listener.h:
```diff
@@ -52,6 +52,9 @@ class agent_ucc_listener : public beerocks_ucc_listener {
      */
     void reset_completed() { m_reset_completed = true; }
+
+    bool has_received_dev_set_config() { return m_received_dev_set_config; }
```
Please add doxygen documentation. 
You could use the same commit messages you've already written to explain when and why this function is used.

vitalii-komisarenko replied:
Done

mariomaz commented on agent/src/beerocks/slave/agent_ucc_listener.h:
```diff
@@ -73,8 +76,9 @@ class agent_ucc_listener : public beerocks_ucc_listener {
-    bool m_reset_completed = false;
+    bool m_in_reset                = false;
+    bool m_reset_completed         = false;
+    bool m_received_dev_set_config = false;
```
Please add doxygen documentation. 
You could use the same commit messages you've already written to explain when and why this function is used.

vitalii-komisarenko replied:
Done

mariomaz reviewed (approved):
Some comments but I let them to your consideration, so approving.
